lib/ostruct.rb

```diff
@@ -221,8 +221,14 @@ def marshal_dump # :nodoc:
   #
   def new_ostruct_member!(name) # :nodoc:
     unless @table.key?(name) || is_method_protected!(name)
-      define_singleton_method!(name) { @table[name] }
-      define_singleton_method!("#{name}=") {|x| @table[name] = x}
+      getter_proc = Proc.new { @table[name] }
+      setter_proc = Proc.new {|x| @table[name] = x}
+      if defined?(::Ractor)
+        ::Ractor.make_shareable(getter_proc)
+        ::Ractor.make_shareable(setter_proc)
```

Comment:
I was thinking this would have the side effect of freezing the OpenStruct instance and `@table` but actually it does not, which is a CRuby bug: https://bugs.ruby-lang.org/issues/18243
A Ractor should never be able to mutate objects of another Ractor directly of course.